### PR TITLE
matmul+activation fuse pass

### PR DESCRIPTION
**Pull request by Silv3S**
### PR types
Performance optimization

### PR changes
OPs

### Describe
- fuse matmul_v1 with all supported activations,
- add required unit tests.

### CODE REVIEW

File: paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc
Comment (jakpiase) [inline paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc]: Why there is no AddOpCompat for `clip` activation function?

File: paddle/fluid/operators/activation_op.cc
Comment (jakpiase) [inline paddle/fluid/operators/activation_op.cc]: Are we allowed to just delete deprecated parameters? I was thinking that due to backward compatibility they must stay as they were before

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (jakpiase) [inline paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: What do you think about creating a common static function which would append eltwises into post_ops object? That behavior is common between all operators that are fused with activation, and all these functions look exactly the same. By using single function we will be able to add support for more activation with less redundancy

**Review by jakpiase — COMMENTED** [review]
Awesome initiative! I have left some comments, but the code looks generally really nice

File: paddle/fluid/operators/activation_op.cc
Comment (Silv3S) [reply paddle/fluid/operators/activation_op.cc]: Currently, there is a discussion and looking for a way to get rid of oneDNN-only attributes, which were used in fuse passes. Since we can get their values via `ExecutionContext`, we could slim base operators down https://github.com/PaddlePaddle/Paddle/issues/43232

File: paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc
Comment (Silv3S) [reply paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc]: I missed it, thanks!

File: paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc
Comment (Silv3S) [reply paddle/fluid/operators/mkldnn/matmul_mkldnn_op.cc]: That's great idea. I will extend previously added `AcquireActivationAlgorithm` to append eltwises into post_ops, instead of just returning `dnnl::algorithm`

**Comment by Silv3S** [comment]
@piotrekobi, @tsocha please review

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc
Comment (tsocha) [inline paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc]: `!` is easy to miss, lets make a helper variable
```suggestion
    auto is_fuse_activation_not_empty = !ctx.Attr<std::string>("fuse_activation").empty();
    if (is_fuse_activation_not_empty) {
```

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc
Comment (tsocha) [inline paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc]: Do you need to pass `activation_scale` as an argument?
I can see that in this context its value is 1.0 or read from attribute.

File: paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h
Comment (tsocha) [inline paddle/fluid/operators/elementwise/mkldnn/elementwise_mkldnn_op.h]: I think that `AppendActivation` can check if `ctx` have a valid `fuse_activation` attribute inside, it will simplify usage of this function(no need to write such `ifs` here and there.

File: paddle/fluid/framework/ir/mkldnn/softplus_activation_mkldnn_fuse_pass.h
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/softplus_activation_mkldnn_fuse_pass.h]: Since its only a variable name you can postpone this change to another PR to reduce number of modified files to 20(no approval needed).

File: paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc:93]: It would be nice to move this if inside `PrettyLogDetail` but I won't enforce it.

File: paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc:56]: Do we need this if here?
It can disable this fuse if this method will complain about custom attrs etc.

File: paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc]: elt_act_mkldnn_fuse_pass don't need these attributes to be set explicitly so this pass don't need it either IMHO.
```suggestion
```

**Review by tsocha — CHANGES_REQUESTED** [review]

File: paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc
Comment (piotrekobi) [inline paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc]: This code is repeated in softplus_activation_mkldnn_fuse_pass.cc and it's probably a good idea to minimize the use of nested if statements. Maybe attr_map can be a created somewhere as a static object instead of being created dynamically during every pass?

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc
Comment (Silv3S) [reply paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc]: Yes, it's needed because conv operator is different than others. All other op+act fuse passes use default value 1.0, only `conv_mkldnn_op` computes it from attributes(`force_fp32_output`, `scale_out` and `activation_scale`).  If more ops compute `activation_scale` I will move this logic to `AppendActivation`.

File: paddle/fluid/framework/ir/mkldnn/softplus_activation_mkldnn_fuse_pass.h
Comment (Silv3S) [reply paddle/fluid/framework/ir/mkldnn/softplus_activation_mkldnn_fuse_pass.h]: That's good idea, but I need 2 other approvals (setting TIMEOUT and removing attrs) anyway, so exceeding 20 files shouldn't delay PR merge

File: paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc
Comment (Silv3S) [reply paddle/fluid/framework/ir/mkldnn/matmul_activation_mkldnn_fuse_pass.cc:56]: I think this topic needs further discussion, because some fuse passes have `OpCompat` for each fusion combination and others don't have any. We should decide if it gives us any benefit and then unify all passes

File: paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc
Comment (Silv3S) [reply paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc]: Great idea 👍 I moved setting `attr_map` to helper function and also created method returning list of all supported activations, so it will be much easier to add new activations to all passes at once.

**Review by jakpiase — DISMISSED** [review]
LGTM

File: paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc
Comment (tsocha) [inline paddle/fluid/framework/ir/mkldnn/elt_act_mkldnn_fuse_pass.cc]: You can reduce number of parameters by moving attr_map extraction inside FuseElementwiseAct function.
Let's do this here and in other passes.
```suggestion
      FuseElementwiseAct(graph, elt_type, act_type);
```

**Review by tsocha — CHANGES_REQUESTED** [review]

File: paddle/fluid/platform/mkldnn_reuse.h
Comment (tsocha) [inline paddle/fluid/platform/mkldnn_reuse.h]: Could you move this if to `AppendActivation` as well? part 2/2
```suggestion
  auto invalid_attribute = ctx.HasAttr("fuse_activation")
                           ? ctx.Attr<std::string>("fuse_activation").empty()
                           : true;
  if (invalid_attribute) return;
```

File: paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc
Comment (tsocha) [inline paddle/fluid/operators/mkldnn/conv_mkldnn_op.cc]: Could you move this if to `AppendActivation` as well? part 1/2
```suggestion
    paddle::platform::AppendActivation(
        ctx, post_operations, activation_scale);
```

**Review by tsocha — CHANGES_REQUESTED** [review]

**Comment by Silv3S** [comment]
@jiangjiajun please review

**Comment by Silv3S** [comment]
@XiaoguangHu01 could you please approve exceeding the limit of modified files?

**Comment by Silv3S** [comment]
@chalsliu could you please approve setting TIMEOUT for newly added unit test?

**Review by tsocha — DISMISSED** [review]

**Review by jczaja — DISMISSED** [review]
LGTM

**Review by piotrekobi — DISMISSED** [review]
LGTM

**Review by jczaja — APPROVED** [review]

**Pull request merged**
